Update Data Pipeline vignette (#1625)
Gilead-BioStats/gsm: vignettes/DataPipeline.Rmd

1.  `dfInput`: Input data; Cross-domain participant-level input data with all needed data for KRI derivation. Created by the `Input_Rate()` function as needed.
2.  `dfTransformed`: Transformed data; Site-level transformed data including KRI calculation. Created by `Transform_*()` functions.
3.  `dfAnalyzed`: Analyzed data; Site-level analysis results data. Created by `Analyze_*()` functions.
4.  `dfFlagged`: Flagged data; Site-level analysis results with flags added to indicate potential statistical outliers. Created by passing numeric thresholds to a `Flag_*()` function.
5.  `dfSummary`: Summary data; Standardized subset of the flagged data. This summary data has the same structure for all assessments and always includes both KRI and Flag values so that a user can easily look at trends for any given site across multiple assessments. Created using the `Summarize()` function.

> Each KRI has a `.yaml` file that sequentially executes all of the supporting functions and returns a list of datasets and visualizations per the specifications of the .yaml file. In the default `yaml` files included in the `workflow/metrics` directory of this package, the 5 standard datasets that are always returned are `dfTransformed`, `dfAnalyzed`, `dfFlagged`, `dfBounds`, and `dfSummary`. A step-by-step example of the data assessment workflow can be found in the [Data Analysis Vignette](https://gilead-biostats.github.io/gsm/articles/DataAnalysis.html).

Reviewer: ```suggestion
Each KRI has a `.yaml` file that sequentially executes all of the supporting functions and returns a list of datasets and visualizations per the specifications of the .yaml file. In the default `yaml` files included in the `system.file("workflow/metrics", package = "gsm")` directory of this package, the 5 standard datasets that are always returned are `dfTransformed`, `dfAnalyzed`, `dfFlagged`, `dfBounds`, and `dfSummary`. A step-by-step example of the data assessment workflow can be found in the [Data Analysis Vignette](DataAnalysis.html).
```

## Appendix 2 - Analysis Workflow Specifications

Assessment workflow metadata objects are passed to the `lWorkflow` parameter in `RunWorkflow()` to define functions and parameters across multiple studies.

The `lWorkflow` object is a named list of metadata and steps defining how each assessment should be run. By default, `MakeWorkflowList()` imports YAML specifications from `workflow/metrics`. Each item in `lWorkflow` expects the following parameters in the `steps` section:

-   `workflow`: Array defining one or more functions to be executed as part of the workflow for a given assessment
    -   `workflow[]$meta`: specifies all of the metadata information for the KRI.
    -   `workflow[]$steps`: specifies all of the steps in the workflow.
        -   `workflow[]$steps$name`: name of the `{gsm}` function.
        -   `workflow[]$steps$inputs`: specifies the required input data
        -   `workflow[]$steps$output`: specifies the output data from the workflow step, which can be used as an input in the next step in the workflow
        -   `workflow[]$steps$params`: specifies parameters to be passed to the function

> For example, the default workflow for the AE assessment (`workflow/metrics/kri0001.yaml`) is shown below:

Reviewer: ```suggestion
For example, the default workflow for the AE assessment (`system.file("workflow/metrics/kri0001.yaml", package = "gsm")` is shown below:
```

The default KRI data pipeline begins with one or more clinical datasets related to the KRI. An optional participant-level subset is then applied to each dataset before aggregating by participant to quantify the KRI at the participant level. This aggregated dataset is then further summarized by site to obtain site-level metrics.

## Generating Input Data

Generating input data is outside the scope of the assessment pipeline. The specifications for input data (i.e., `dfInput`) are designed to be flexible enough so that this data frame can be generated from multiple clinical data standards (e.g., Raw, rawplus, or CDISC-compliant ADaM). In order to generate input data, the user uses the data frames outputted by the `data_mappings.yaml` file along with the `Input_Rate()` function to generate the `dfInput` needed for a particular KRI.

> This `data_mapping.yaml` file is provided for convenience, but may not work for all clinical studies. When no relevant data.frame is produced with the `data_mapping.yaml`, the user can use the `RunQuery()` function to write a SQL query that produces the properly mapped and filtered dataframe. Details on this can be found in the [Cookbook Vignette](https://github.com/Gilead-BioStats/gsm/wiki/Cookbook-Vignette).

Author: for me, this just goes to the Cookbook-Vignette page on the gsm wiki, which will need to be updated with the new cookbook (and all other vignettes) at some point, so thanks for catching that.

> This `data_mapping.yaml` file is provided for convenience, but may not work for all clinical studies. When no relevant data.frame is produced with the `data_mapping.yaml`, the user can use the `RunQuery()` function to write a SQL query that produces the properly mapped and filtered dataframe. Details on this can be found in the [Cookbook Vignette](https://github.com/Gilead-BioStats/gsm/wiki/Cookbook-Vignette).

Reviewer: `devtools::install(build_vignettes = TRUE)` results in the proper link to a local copy of the vignette, and that blows my mind 🙃

> This `data_mapping.yaml` file is provided for convenience, but may not work for all clinical studies. When no relevant data.frame is produced with the `data_mapping.yaml`, the user can use the `RunQuery()` function to write a SQL query that produces the properly mapped and filtered dataframe. Details on this can be found in the [Cookbook Vignette](https://github.com/Gilead-BioStats/gsm/wiki/Cookbook-Vignette).

Reviewer: I *think* this should still be updated to point to Cookbook.html directly.

```suggestion
This `data_mapping.yaml` file is provided for convenience, but may not work for all clinical studies. When no relevant data.frame is produced with the `data_mapping.yaml`, the user can use the `RunQuery()` function to write a SQL query that produces the properly mapped and filtered dataframe. Details on this can be found in the [Cookbook Vignette](Cookbook.html).
```

## Analysis Data Pipeline

![](data-analysis.PNG){width="100%"}

The image above provides an overview of the default KRI analysis pipeline. The pipeline is a standardized five-step process for assessing data issues by going from participant-level input data to a standardized site-level summary of model results. The functions used in each step of the data pipeline along with the input and output datasets are described in more detail below.

1.  `dfInput`: Input data; Cross-domain participant-level input data with all needed data for KRI derivation. Created by the `Input_Rate()` function as needed.

> 2.  `dfTransformed`: Transformed data; Site-level transformed data including KRI calculation. Created by `Transform()` functions.

Reviewer: ```suggestion
2.  `dfTransformed`: Transformed data; Site-level transformed data including KRI calculation. Created by `Transform_*()` functions.
```

Standardizing references to function families.

> 3.  `dfAnalyzed`: Analyzed data; Site-level analysis results data. Created by `Analyze()` functions.

Reviewer: ```suggestion
3.  `dfAnalyzed`: Analyzed data; Site-level analysis results data. Created by `Analyze_*()` functions.
```

Standardizing references to function families.

> 4.  `dfFlagged`: Flagged data; Site-level analysis results with flags added to indicate potential statistical outliers. Created by passing numeric thresholds to a `Flag()` function.

Reviewer: ```suggestion
4.  `dfFlagged`: Flagged data; Site-level analysis results with flags added to indicate potential statistical outliers. Created by passing numeric thresholds to a `Flag_*()` function.
```

Standardizing references to function families.

Note that each step in this data pipeline can be customized based on the requirements for a specific KRI. The graphic below shows four such workflows.

![](data-analysis-combined.PNG){width="100%"}

## Reporting Data Pipeline

The pipeline discussed here is a standardized process for creating datasets by going from a standardized site-level summary of model results and external data sources to a series of datasets that provide all group-level data and metadata needed to create charts and reports for a given KRI. Unlike the analysis workflow, there is a single workflow for reporting, as all metric-specific information was captured in the analysis phase of the overall pipeline. The functions used in each step of the data pipeline along with the input and output datasets are described in more detail below.

1. `dfSummary`: Summary data; Standardized subset of the flagged data. This summary data has the same structure for all assessments and always includes both KRI and Flag values so that a user can easily look at trends for any given site across multiple assessments. Created using the `Summarize()` function.

> 4. `dfFlagged`: Flagged data; Site-level analysis results with flags added to indicate potential statistical outliers. Created by passing numeric thresholds to a `Flag()` function.

Reviewer: ```suggestion
4. `dfFlagged`: Flagged data; Site-level analysis results with flags added to indicate potential statistical outliers. Created by passing numeric thresholds to a `Flag_*()` function.
```

Standardizing references to function families.

Once all reporting data.frames are created using the `data_reporting.yaml` workflow as discussed above, a user will create charts for all of the metrics specified in the analysis pipeline. `Visualize_Metric()` is a function creates all available charts for a metric using the data provided. If the `dfSummary` provided to `Visualize_Metric()` contains multiple data with multiple dates, then the `timeSeries` charts will be included in the list of charts returned. The output of this function is a list containing the following charts:

 - `scatterJS`: A scatter plot using JavaScript.
 - `scatter`: A scatter plot using ggplot2.
 - `barMetricJS`: A bar chart using JavaScript with metric on the y-axis.
 - `barScoreJS`: A bar chart using JavaScript with score on the y-axis.
 - `barMetric`: A bar chart using ggplot2 with metric on the y-axis.
 - `barScore`: A bar chart using ggplot2 with score on the y-axis.

>  - `timeSeriesContinuousScoreJS`: A time series chart using JavaScript with score on the y-axis.
>  - `timeSeriesContinuousMetricJS`: A time series chart using JavaScript with metric on the y-axis.

Reviewer: ```suggestion
 - `timeSeriesContinuousMetricJS`: A time series chart using JavaScript with metric on the y-axis.
 - `timeSeriesContinuousScoreJS`: A time series chart using JavaScript with score on the y-axis.
```
Previous charts were in "metric then score" order.

 - `timeSeriesContinuousNumeratorJS`: A time series chart using JavaScript with numerator on the y-axis.

Now that all imputs for the `Report_KRI()` function have been produced, it can be called to create a detailed report showing both charts and listings summarizing each KRI that was run for the study.

Putting all of these steps together, the following script will create a sample report using both the analysis and reporting workflows for the AE KRI.

```{r, eval = FALSE}
lRaw <- list(
    dfSUBJ = clindata::rawplus_dm,
    dfAE = clindata::rawplus_ae,
    dfPD = clindata::ctms_protdev,
    dfLB = clindata::rawplus_lb,
    dfSTUDCOMP = clindata::rawplus_studcomp,
    dfSDRGCOMP = clindata::rawplus_sdrgcomp %>% dplyr::filter(.data$phase == 'Blinded Study Drug Completion'),
    dfDATACHG = clindata::edc_data_points,
    dfDATAENT = clindata::edc_data_pages,
    dfQUERY = clindata::edc_queries,
    dfENROLL = clindata::rawplus_enroll
)

> #make workflow lists

Reviewer: ```suggestion
# Make Workflow Lists
```

wf_kri <- MakeWorkflowList(strNames="kri0001")
wf_reporting <- MakeWorkflowList(strNames = "reporting")

# Generate Mapped Data
lMapped <- RunWorkflows(lWorkflow = wf_mapping, lData = lRaw)

# Generate Analysis Results Data
lAnalysis <- RunWorkflows(lWorkflow = wf_kri, lData = lMapped)

# Generate Reporting Data
lReporting_Input <- list(
    ctms_site = clindata::ctms_site,
    ctms_study = clindata::ctms_study,
    dfEnrolled =lMapped$dfEnrolled,
    lWorkflows = wf_kri,
    lAnalysis = list(kri0001 = lAnalysis),
    dSnapshotDate = Sys.Date(),
    strStudyID = "ABC-123"
)

lReporting <- RunWorkflows(lWorkflow = wf_reporting, lData = lReporting_Input)

# Convenience Mappings
dfGroups <- lReporting$dfGroups
dfMetrics <- lReporting$dfMetrics
dfSummary <- lReporting$dfSummary
dfBounds <- lReporting$dfBounds

# Create dfSite and dfStudy for charts
dfSite <- dfGroups %>%
  filter(GroupLevel == "Site") %>%
  pivot_wider(names_from=Param, values_from=Value) %>%
  rename(
    SiteID = GroupID,
    status = Status,
    enrolled_participants = ParticipantCount
  )

dfStudy <-  dfGroups %>% 
  filter(GroupLevel == "Study") %>% 
  pivot_wider(names_from=Param, values_from=Value)

> #Create Charts for all metrics

Reviewer: ```suggestion
# Create Charts for all metrics
```

options(vsc.viewer = FALSE)
metrics <- unique(dfMetrics$MetricID)
charts <- metrics %>% map(~Visualize_Metric(
  dfSummary = dfSummary,
  dfBounds = dfBounds,
  dfGroups = dfGroups,
  dfMetrics = dfMetrics,
  strMetricID = .x

> )
> ) %>% setNames(metrics)

Reviewer: ```suggestion
)) %>% setNames(metrics)
```
Since the parentheses are opened on a single line, it's less confusing if they're also closed on a single line.

##function not currently working!
# Report_KRI(lCharts = charts,
#            dfSummary = dfSummary, 
#            dfSite = dfSite,
#            dfStudy = dfStudy,
#            dfMetrics = dfMetrics,
#            strOutput = "~/test_report_kri.html")
```

> Additional examples are provided in the [Cookbook Vignette](https://gilead-biostats.github.io/gsm/articles/Cookbook.html).

Reviewer: ```suggestion
Additional examples are provided in the [Cookbook Vignette](Cookbook.html).
```
Or link in the same way as the previous one, if the magic to recode the link is intentional.

# Appendix 1 - Data Model

# Overview 
![](data-model.PNG){width="100%"}

# Analytics data model

The KRI analytics pipeline is a standardized process for **Analyzing** data issues by going from participant-level `input` data to a standardized site-level `summary` of model results. The data sets used in each step of the data pipeline are described in detail below.

## Analytics Data Tables
  
### `dfInput`
  - Function(s) used to create table: 
    - `Input_Rate()`
  - Inputs: 
    - `dfSubjects`
    - `dfNumerator`
    - `dfDenominator`

>   - Usage: The base data.frame for all Analytics workflows. Feeds into the `Transform_XX` functions.

Reviewer: ```suggestion
  - Usage: The base data.frame for all Analytics workflows. Feeds into the `Transform_*()` functions.
```
Standardizing references to function families.

  - Structure:

  | Table    | Column Name  | Description                          | Type     | Optional  | 
  |----------|--------------|--------------------------------------|----------|--|
  | dfInput  | SubjectID    | The subject ID                       | Character| |
  | dfInput  | GroupID      | The group ID for the metric          | Character| | |
  | dfInput  | GroupLevel   | The group type for the metric (e.g. "Site")    | Character| | 
  | dfInput  | Numerator    | The calculated numerator value       | Numeric  | |
  | dfInput  | Denominator  | The calculated denominator value     | Numeric | |
  | dfInput  | Metric       | The calculated rate/metric value     | Numeric  | |

### `dfTransformed`
  - Function(s) used to create table:
    - `Transform_Rate()`
    - `Transform_Count()`
  - Inputs: `dfInput`

>   - Usage: Convert from input data format to needed format to derive KRI for an Assessment via the `Analyze_XX` functions.

Reviewer: ```suggestion
  - Usage: Convert from input data format to needed format to derive KRI for an Assessment via the `Analyze_*()` functions.
```

  - Structure:

  | Table    | Column Name  | Description                          | Type     | Optional  | 
  |----------|--------------|--------------------------------------|----------|--|
  | dfTransformed | GroupID      | The group ID for the metric          | Character| | |
  | dfTransformed | GroupLevel   | The group type for the metric (e.g. "Site")    | Character| | 
  | dfTransformed | Numerator    | The calculated numerator value       | Numeric  | |
  | dfTransformed | Denominator  | The calculated denominator value     | Numeric | |
  | dfTransformed | Metric       | The calculated rate/metric value     | Numeric  | |

###  `dfAnalyzed`
  - Function(s) used to create table:
    - `Analyze_Fisher()`
    - `Analyze_Identity()`
    - `Analyze_NormalApprox()`
    - `Analyze_Poisson()`
    - `Analyze_QTL()`
  - Inputs: `dfTransformed`

>   - Usage: Prepare the data for `Flag_XX` by performing the specified test on the metric provided.

Reviewer: ```suggestion
  - Usage: Prepare the data for `Flag_*()` by performing the specified test on the metric provided.
```

|-------------|----------------------|-----------------------------------|----------|
| Study       | Status               | Study Status                      | Character| 
| Study       | Title       | Protocol Title       | Numeric | 
| Study       | ParticipantCount       | # of Enrolled Participants        | Numeric | 
| Study       | SiteCount              | # of Activated Sites                 | Numeric| 
| Study       | ParticipantsPlanned        | # of Planned Participants        | Numeric| 
| Study       | SitesPlanned               | # of Planned Sites        | Numeric| 
| Site       | ParticipantCount       | # of Enrolled Participants        | Numeric | 
| Site       | Status       | Site Status       | Character | 
| Site       | InvestigatorFirstName       | Investigator First name        | Character | 
| Site       | InvestigatorLastName       | Investigator Last name        | Character | 
| Site       | City        | City       | Character| 
| Site       | State       | State      | Character | 
| Site       | Country       | Country       | Character | 
| Country       | EnrolledParticipants       | # of Enrolled Participants        | Numeric | 

###  `dfMetrics`

>   - Function used to create table:

Reviewer: It looks like these sections from here to the end are incomplete.

| dfStudy| act_lplv              | Actual last patient last visit       | Date   | |
| dfStudy| est_lpfv              | Estimated last patient first visit     | Date   | |
| dfStudy| act_lpfv              | Actual last patient first visit       | Date   | |
| dfStudy| status                | Study Status                       | Character| |
| dfStudy| num_plan_subj         | Number of planned subjects in study      | Numeric | |
| dfStudy| num_enrolled_subj_m   | Number of enrolled subjects in study | Numeric | |
| dfStudy| protocol_indication   | Protocol Indication                        | Character| |
| dfStudy| product               | Product                      | Character| |
| dfStudy| therapeutic_area      | Therapeutic Area                 | Character| |
| dfStudy| protocol_product_number| Protocol Product Number                      | Numeric | |
| dfStudy| x_rbm_flg             | Is RBM Flagged                      | Character| |

## Appendix 2 - Analysis Workflow Specifications

Assessment workflow metadata objects are passed to the `lWorkflow` parameter in `RunWorkflow()` to define functions and parameters across multiple studies.

> The `lWorkflow` object is a named list of metadata and steps defining how each assessment should be run. By default, `MakeWorkflowList()` imports YAML specifications from `inst/workflow/metrics`. Each item in `lWorkflow` expects the following parameters in the `steps` section:

Reviewer: See previous discussion of `inst`.

-   `workflow`: Array defining one or more functions to be executed as part of the workflow for a given assessment
    -   `workflow[]$meta`: specifies all of the metadata information for the KRI.
    -   `workflow[]$steps`: specifies all of the steps in the workflow.
        -   `workflow[]$steps$name`: name of the `{gsm}` function.
        -   `workflow[]$steps$inputs`: specifies the required input data
        -   `workflow[]$steps$output`: specifies the output data from the workflow step, which can be used as an input in the next step in the workflow
        -   `workflow[]$steps$params`: specifies parameters to be passed to the function

> For example, the default workflow for the AE assessment (`inst/workflow/metrics/kri0001.yaml`) is shown below:

Reviewer: Se previous discussion of `inst`

  strMetricID = .x
)) %>% setNames(metrics)

> #function not currently working!

Reviewer: Should we mention an issue related to this? I want to make sure this isn't being left in accidentally (vs intentionally).

###  `dfMetrics`

>   - Workflow used to create table: 

Reviewer: Is this intentionally blank?
